### PR TITLE
Fix the serialization of RetinaNet

### PR DESCRIPTION
**Pull request by haifeng-jin**
removed the `anchor_generator` from `get_config()` because it cannot be provided together with `label_encoder` and `prediction_decoder` while initializing the class.

I suggest we start to add these serialization tests when implementing new models.
Otherwise, many models may not work in saving.

### CODE REVIEW

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py [hunk 1]
```diff
@@ -64,9 +64,22 @@ def call(self, x, training=False):
...
+    @classmethod
+    def from_config(cls, config):
+        config.update(
+            {
+                "bias_initializer": keras.initializers.deserialize(
+                    config["bias_initializer"]
+                )
+            }
+        )
+        return super().from_config(config)
```
Comment (haifeng-jin) [inline keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:85]: @nkovela1 Would you please take a look if I did something weird here? Is there a way to implicitly serialize and deserialize the initializer? Thanks.

Comment (nkovela1) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:85]: Hi Haifeng,

I believe the explicit serialization and deserialization of the bias initializer is unnecessary here. It should already be taken care of unless you are using a custom (non-Keras) initializer. Please let me know if removing serialization there results in any errors.

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py [hunk 2]
```diff
@@ -64,9 +64,22 @@ def call(self, x, training=False):
 
     def get_config(self):
         config = {
-            "bias_initializer": self.bias_initializer,
+            "bias_initializer": keras.initializers.serialize(
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:67]: Does this mean that we won't be able to serialize the string "zeros", and the actual keras initializer will have to be passed?

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py (hunk 1 above)
Comment (haifeng-jin) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:85]: We will stick to the explicit serialization for initializers for now : )
Thanks, @nkovela1!

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 3]
```diff
@@ -516,12 +518,9 @@ def get_config(self):
         return {
             "num_classes": self.num_classes,
             "bounding_box_format": self.bounding_box_format,
-            # TODO(bischof): actually serialize the backbone
             "backbone": self.backbone,
-            "anchor_generator": self.anchor_generator,
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:521]: Won't removing this break serialization when a custom anchor_generator is used?

**Review by ianstenbit — COMMENTED** [review]
A few questions -- sorry I'm a bit of a rookie when it comes to this serialization stuff!

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py (hunk 2 above)
Comment (haifeng-jin) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:67]: Yes. I see `PredictionHead` as a internal only class, so we may enforce it. If we still want the "zeros", we  can do a special case here for the string types.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 3 above)
Comment (haifeng-jin) [reply keras_cv/models/object_detection/retina_net/retina_net.py:521]: Yes, this is a problem. I am not so sure about the logics between `anchor_generator`, `label_encoder`, `prediction_decoder`. When `get_config()` is called, they all exist. However, they cannot be all passed to `__init__()`. What would be a good way to judge which one to omit here?

Comment (ianstenbit) [reply keras_cv/models/object_detection/retina_net/retina_net.py:521]: Okay actually I think what you've done is exactly right -- since the encoder/decoder are dependent on this, just serializing the encoder/decoder should be good.

This makes me wonder though -- @LukeWood why can't we pass a custom `anchor_generator` in addition to a custom `prediction_decoder`? It seems like those are now decoupled, so I wonder if this is an artifact of how this used to be structured?

**Review by ianstenbit — APPROVED** [review]
Thanks! This LGTM -- I left one question for @LukeWood that we can probably handle in a follow-up

**Comment by haifeng-jin** [comment]
\gcbrun

**Comment by ianstenbit** [comment]
> \gcbrun

Gotta use a forward-slash 😄

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:521]: > Yes, this is a problem. I am not so sure about the logics between anchor_generator, label_encoder, prediction_decoder. When get_config() is called, they all exist. However, they cannot be all passed to __init__(). What would be a good way to judge which one to omit here?

We definitely should not  prevent the use of custom anchor generators.  One thing we could do is exclude anchor generators, and encourage only the use of `save_weights()` and `load_weights()`?  I'm not sure the right approach, but anchor generators and weights are tightly coupled unfortunately.

**Comment by LukeWood** [comment]
Realistically, this needs a deeper dive (probably after I finish forward pass testing). 

I think this is a step in the right direction, so lets go for it with the note to revisit once we have strong numberics.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 3 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/retina_net/retina_net.py:521]: @LukeWood -- since the label encoder contains the anchor generator, by serializing the label encoder aren't we also effectively serializing the anchor generator?

I feel like I'm missing something. Sorry if it's a stupid question!

**Comment by haifeng-jin** [comment]
@LukeWood Was this PR somehow pending on https://github.com/keras-team/keras-cv/pull/1458? Just add this note for future reference.
